### ISSUE TITLE
Oracle.ManagedDataAccess.Core new OracleConnection instances appear to create new OraclePoolManager instances that are never released

### ISSUE THREAD

**Issue by BashouT** [oracle/dotnet-db-samples#196]
`Oracle.ManagedDataAccess.Core` versions 3.21.1, 2.19.110 and 2.19.101.

It's possible I am missing something/misunderstanding here and hopefully you could fill me in. (removed some content for data protection/brevity)

During each request to our API where needed we create a new connection as credentials are supplied via the caller of the API. This is done through a factory where we return a new instance of our "DirectDbConnection" object. 

Factory Code:

    OracleCredential oracleCredentials = GetCurrentCredentials();
    var ora = new OracleConnection(_connectionStrings.SocialCareDirectConnectionString, oracleCredentials);
    var DirectDatabaseConnection = new MosaicDirectDatabaseConnection(ora, _logger, _correlationIdService);
    return DirectDatabaseConnection;

Direct Connection Code (removed logging and exception handling code):

    public DirectDatabaseConnection(OracleConnection oracleConnection)
    {
        _oracleConnection = oracleConnection;
    }
        
    public void Dispose()
    {
        OracleConnection.ClearPool(_oracleConnection);
        _oracleConnection?.Close();
        _oracleConnection?.Dispose();
    }

    public IEnumerable<T> Query<T>(string query, bool buffered, int commandTimeout)
    {
        IEnumerable<T> result = new List<T>();

        if (_oracleConnection.State == ConnectionState.Closed)
        {
            _oracleConnection.Open();
        }

        if (_oracleConnection.State == ConnectionState.Open)
        {

                result = _oracleConnection.Query<T>(query, buffered: buffered, commandTimeout: commandTimeout);
        }
        return result;
    }

Our typical connection string looks like this and is pulled from config that is DI'd to the factory:
`Data Source=(DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=)(PORT=))(CONNECT_DATA=(SERVICE_NAME=)));`

The connection is then used like so:

    using var connection = _connectionFactory.GetInstance();
    List<> result = null;
    string query = "'";

    var dbResult = connection.Query<>(query, buffered: true, commandTimeout: 60)
        .ToList();

    result = MapDbResult(dbResult);
    return new List<Actual>(result);

I'm sure that technically this is not an optimal approach in a performance manner, however situation states I need to fix with as little change as possible in the short term.

From my understanding connection pooling by default would only create a new connection pool if the connection string is in any way different. In my testing we've always used the same connection string AND credentials so form a connection pool standpoint it should always be the same if it exists. However the issue that is occurring is that the `OraclePoolManager` instances that are created when we create a new connection (it appears) are not being garbage collected. This led to the application reaching gigabytes of RAM usage overtime as this application is queried a lot by some dependent systems

Since there is no documentation on this internal class from what I can find it's hard to understand why that may be.

Any insight you could shed would be greatly appreciated :)

**Comment by alexkeh** [comment]
This shouldn't be happening. Can you put together a stand-alone test case that will reproduce the leak reliably? We'll take a look.

It can be challenging reproducing from code snippets, especially if there's a code path in which connections are not closed/disposed.

**Comment by alexkeh** [comment]
Will re-open upon receiving a test case.

**Comment by Bashout** [comment]
Apologies for the slow reply:

I've uploaded a simplified equivalent of what I tried to explain with snippets [here](https://github.com/Bashout/OracleDbConnectionLeakSample)

Should be pretty quick to run up and see if you have a similar outcome.

**Comment by alexkeh** [comment]
Thanks for the test case. I filed a bug (33375248) to have one of our dev team members review the test case.

One thing I noticed about your test case is that you are calling ClearPool in your Dispose custom method. Why is that? ClearPool will remove all pooled connections, forcing your pool to then rebuild itself later on.

**Comment by Bashout** [comment]
>Thanks for the test case. I filed a bug (33375248) to have one of our dev team members review the test case.

Thanks for reopening, great to get another set of eyes on this.

> One thing I noticed about your test case is that you are calling ClearPool in your Dispose custom method. Why is that? ClearPool will remove all pooled connections, forcing your pool to then rebuild itself later on.

Yeah, so it's kind of an odd one where sessions, from the DBAs who own the system we interact with, dislike the image of having x or y amount of sessions open even if they're inactive. (We received a number of complaints regarding there being a large number of connections open)

I suppose a better solution would've been to disable connection pooling.

**Comment by alexkeh** [comment]
The Clear Pool calls on every Dispose defeat the whole purpose of using a connection pool. Your app will be destroying and rebuilding the connection pool over and over again.

In fact, that could be the reason the bug occurs for you and no one else has reported it. You end up using the connection pool in a way that is not intended. It wouldn't be a scenario Oracle tests as we wouldn't expect customers to use the pool in this way.

Does the problem occur when you remove the ClearPool from the Dispose?

**Comment by Bashout** [comment]
> The Clear Pool calls on every Dispose defeat the whole purpose of using a connection pool. Your app will be destroying and rebuilding the connection pool over and over again.
I understand this - as I said, disabling pooling would've been a better alternative, however even by specifying `'pooling=false'` in the connection string this issue still occurs.

> Does the problem occur when you remove the ClearPool from the Dispose?
Yes the problem still occurs without this.

**Comment by alexkeh** [comment]
Well, that is odd if the pool manager is activated when Pooling=false. I'll add this info to the bug log.

**Comment by Bashout** [comment]
Just wanted to provide a small update on this from my end. Work wise we're waiting on Oracle guidance in terms of resolving this issue. 

I reduced my test sample locally to the smallest possible solution which would spin up a new connection, open it and then instantly close it and I found that the pool manager is created and maintained long after the connection has been disposed of and left out of scope. Unless I've missed something this seems like a fundamental issue in the DB adaptor.

I appreciate you raising the bug on my behalf 👍, is there anyway I could get a look at any updates on it as it progresses?

**Comment by alexkeh** [comment]
I've published the bug so that you can track it if you have a My Oracle Support (MOS) account. If there's some urgency to having the issue fixed, let us know.

Next month is the .NET 6 and EF Core 6 release, which is what the Oracle .NET dev team is focused on supporting as quickly as possible.

**Comment by alexkeh** [comment]
After investigating the test case, a new OracleCredential object was created every time a new DB connection was created. That was what led to the new PoolManager instances. To limit the creation of all these PoolManager instances, the first OracleCredential object should be reused.

Below is a modification to the test case that checks whether an OracleCredential object has already been created. If so, then it is reused to prevent additional PoolManagers.

```
using Oracle.ManagedDataAccess.Client;
using System;
using System.Collections.Generic;
using System.Security;
using System.Text;

namespace OracleDbConnectionLeak.Db
{
    public class DirectDbConnectionFactory : IDirectDbConnectionFactory
    {
        private const string connectionString = "Data Source=";
        private const string username = "";
        private const string pw = "";
        private static int count = 0;
        private static OracleCredential oracleCredentials;
        private static Boolean credCreated=false;

        public DirectDbConnectionFactory()
        {
        }

        public ICustomDbConnection GetInstance()
        {
            if(credCreated==false)
            {
                oracleCredentials = GetCurrentCredentials();
                credCreated = true;
            }
            

            var ora = new OracleConnection(connectionString, oracleCredentials);

            var mosaicDirectDatabaseConnection = new DbConnection(ora);

            return mosaicDirectDatabaseConnection;
        }

        private OracleCredential GetCurrentCredentials()
        {
            count++;
            Console.WriteLine("count: " + count);

            var securePw = new SecureString();

            foreach (char c in pw)
                securePw.AppendChar(c);

            securePw.MakeReadOnly();

            var oCredential = new OracleCredential(username, securePw);
            return oCredential;
        }
    }
}

**Comment by sungam3r** [comment]
Was the actual OracleCredential value the same every time? Just curious.

**Comment by alexkeh** [comment]
Yes.

**Comment by BashouT** [comment]
@alexkeh why does this occur even if pooling is disabled? Also it seems odd that you should have to maintain the credentials object to reuse when pooling is enabled. Shouldn't the driver identify there's a pool available for that credential pair even if the object is a different reference?

**Comment by alexkeh** [comment]
@BashouT Clearing the pool removes the connections in the pool, but you still need a pool manager to manage the pool, which why it remains around.

Maintaining the same credential object is standard ADO.NET behavior. From the [SqlClient connection pooling](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-connection-pooling) doc:

`Different instances of SqlCredential will use different connection pools, even if the user ID and password are the same.`

In general, ODP.NET tries to stay in line with standard ADO.NET provider behavior as that is what developers expect.

**Comment by BashouT** [comment]
Okay @alexkeh , makes sense in that case. The only question that remains for me is why a pool manager is created when pooling is disabled?

**Comment by sungam3r** [comment]
From the same [docs](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-connection-pooling#pool-fragmentation-due-to-integrated-security):
![изображение](https://user-images.githubusercontent.com/21261007/140403941-0df5c1d4-47ef-48c9-9348-30e1a9e778d4.png)

That is, I join the question from @BashouT.

**Comment by alexkeh** [comment]
The PoolManager manages connections for non-pooled connections as well. In that scenario, it manages tasks, such as honoring "connection timeout" by using a worker thread to create a connection, rather than an application thread, to avoid the application thread hanging indefinitely.

**Comment by sungam3r** [comment]
That is, each `PoolManager` also allocates additional worker thread with its own memory for stack even if pooling is disabled. Right? If I understand this right then in picture from https://github.com/oracle/dotnet-db-samples/issues/196#issuecomment-947117516  we would see 15 additional `System.Threading.Thread` objects in the heap. Right?

**Comment by alexkeh** [comment]
I'm not sure how many additional thread objects one should see.

Since the main question has been resolved, I am closing this thread.

**Comment by mihaimyh** [comment]
How was it resolved, what is the resolution?

**Comment by alexkeh** [comment]
@mihaimyh The resolution was to limit the creation of all these PoolManager instance by reusing the first OracleCredential object. Do not create a new OracleCredential object for the next connection to be opened unless you intend to create a separate, distinct pool.

**Comment by kempcalalo** [comment]
@alexkeh 

Hi, 

But what if I am using a full connection string with username / password and not using a user credential object?
Would the trick to use OracleCredential fix the ConnectionPool? 

For example:

`    public class BaseRepository : IBaseRepository
    {
        public string ConnectionString { get; set; }

        public BaseRepository(IConfiguration configuration)
        {
            ConnectionString = configuration.GetConnectionString("OracleConnection");
            DefaultTypeMap.MatchNamesWithUnderscores = true;
        }
        public IEnumerable<T> Query<T>(string query, object parameters = null, CommandType? commandType = null)
        {
            using OracleConnection conn = new OracleConnection(ConnectionString);

            return commandType.HasValue ? conn.Query<T>(query, parameters, commandType: commandType) : conn.Query<T>(query, parameters);
        }
        public void ChangeDbConnection(string connString)
        {
            ConnectionString = connString;
        }
}`

**Comment by alexkeh** [comment]
@kempcalalo If you are not changing your connection string and not using OracleCredential, then only one pool will be created. In almost all cases, a change to your connection string will result in a new connection pool getting created.

**Comment by kempcalalo** [comment]
> 

@alexkeh 

The thing is we are changing the connection string on the fly, we pull the connection string from a table and then connect to that database and we're not using OracleCredential. Do you suggest using the same technique you mentioned above?

**Comment by alexkeh** [comment]
When you keep changing the connection string, you will create a new pool every time Open() is called. If your intent is to use one pool only, then do not change your connection string.

OracleCredential will not solve your problem of generating more pools. The key is to stop changing the physical string itself.
